Use allclose and relative tolerance in unit testing

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32582 `torch.pow` Add type promotion support and fix issue with __rpow__
* #32581 Simplify precision in type promotion tests.
* **#32538 Use allclose and relative tolerance in unit testing**

Related to issue https://github.com/pytorch/pytorch/issues/31135

In our unit testing we're typically using absolute precision / tolerances to validate results.

Single-precision floating point values hold about 7 decimal digits of precision no matter where they fall, e.g. 123456789f can only be represented as 12345679e8, allowing an absolute tolerance of ~3, but 0.123456789 can handle a tolerance of about ~3e-9. In some cases this requires us to be more conservative than we'd like with our value checking.

By using relative tolerance in assertEqual in unit tests, we should be able to simplify some of our complicated precision handling code described in issue #31135.

nairbv committed:
Use allclose and relative tolerance in unit testing

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit c4a91a59:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c4a91a59a13a6c4f176d2879769c123cf3be563a).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4323917) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89859300) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89859300))
<details>
<summary>
<code>Jan 23 17:08:11 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff</code>
</summary>

```
Jan 23 17:08:11 File to patch:  
Jan 23 17:08:11 Skip this patch? [y]  
Jan 23 17:08:11 4 out of 4 hunks ignored 
Jan 23 17:08:11 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 23 17:08:11 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 23 17:08:11 Traceback (most recent call last): 
Jan 23 17:08:11   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 23 17:08:11     patch_repo(args) 
Jan 23 17:08:11   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 23 17:08:11     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 23 17:08:11 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 23 17:08:11 + cleanup 
Jan 23 17:08:11 + retcode=1 
Jan 23 17:08:11 + set +x 
Jan 23 17:08:11 =================== sccache compilation log =================== 
Jan 23 17:08:11 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 23 17:08:11 Compile requests                 1 
Jan 23 17:08:11 Compile requests executed        0 
Jan 23 17:08:11 Cache hits                       0 
Jan 23 17:08:11 Cache misses                     0 
Jan 23 17:08:11 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

gchanan commented on test/common_utils.py:
```diff
@@ -810,23 +810,16 @@ def assertTensorsEqual(a, b):
-                            diff = diff.abs()
-                        max_err = diff.max()
-                        self.assertLessEqual(max_err, prec, message)
+                            self.assertTrue(torch.allclose(a, b, atol=prec, rtol=prec, equal_nan=True), msg=message)
```
using `prec` for atol and rtol seems surprising?

gchanan reviewed (commented):
I'd want to see an example where this simplifies things

nairbv commented:
> I'd want to see an example where this simplifies things

Added two more commits on the stack, neither of which pass without this change. I think there's quite a bit of testing code we could simplify over time.

I also don't love using the same `prec` for both atol and rtol, but would rather not change all call-sites, and non-zero atol is often still valuable. On master prec means atol, and rtol is left as default -- I think this is a (small) improvement to what we currently have.

nairbv committed:
Update on "Use allclose and relative tolerance in unit testing"


Related to issue https://github.com/pytorch/pytorch/issues/31135

In our unit testing we're typically using absolute tolerances to validate results.

Single-precision floating point values hold about 7 decimal digits of precision no matter where they fall, e.g. 123456789f can only be represented as 12345679e8, allowing an absolute tolerance of ~3, but 0.123456789 can handle a tolerance of about ~3e-9. In some cases this requires us to be more conservative than we'd like with our value checking.

By using relative tolerance in assertEqual in unit tests, we should be able to simplify some of our complicated precision handling code described in issue #31135.


[ghstack-poisoned]

nairbv committed:
Update on "Use allclose and relative tolerance in unit testing"


Related to issue https://github.com/pytorch/pytorch/issues/31135

In our unit testing we're typically using absolute tolerances to validate results.

Single-precision floating point values hold about 7 decimal digits of precision no matter where they fall, e.g. 123456789f can only be represented as 12345679e8, allowing an absolute tolerance of ~3, but 0.123456789 can handle a tolerance of about ~3e-9. In some cases this requires us to be more conservative than we'd like with our value checking.

By using relative tolerance in assertEqual in unit tests, we should be able to simplify some of our complicated precision handling code described in issue #31135.


[ghstack-poisoned]

nairbv committed:
Update on "Use allclose and relative tolerance in unit testing"


Related to issue https://github.com/pytorch/pytorch/issues/31135

In our unit testing we're typically using absolute tolerances to validate results.

Single-precision floating point values hold about 7 decimal digits of precision no matter where they fall, e.g. 123456789f can only be represented as 12345679e8, allowing an absolute tolerance of ~3, but 0.123456789 can handle a tolerance of about ~3e-9. In some cases this requires us to be more conservative than we'd like with our value checking.

By using relative tolerance in assertEqual in unit tests, we should be able to simplify some of our complicated precision handling code described in issue #31135.


[ghstack-poisoned]

nairbv committed:
Update on "Use allclose and relative tolerance in unit testing"


Related to issue https://github.com/pytorch/pytorch/issues/31135

In our unit testing we're typically using absolute tolerances to validate results.

Single-precision floating point values hold about 7 decimal digits of precision no matter where they fall, e.g. 123456789f can only be represented as 12345679e8, allowing an absolute tolerance of ~3, but 0.123456789 can handle a tolerance of about ~3e-9. In some cases this requires us to be more conservative than we'd like with our value checking.

By using relative tolerance in assertEqual in unit tests, we should be able to simplify some of our complicated precision handling code described in issue #31135.


[ghstack-poisoned]

nairbv committed:
Update on "Use allclose and relative tolerance in unit testing"


Related to issue https://github.com/pytorch/pytorch/issues/31135

In our unit testing we're typically using absolute tolerances to validate results.

Single-precision floating point values hold about 7 decimal digits of precision no matter where they fall, e.g. 123456789f can only be represented as 12345679e8, allowing an absolute tolerance of ~3, but 0.123456789 can handle a tolerance of about ~3e-9. In some cases this requires us to be more conservative than we'd like with our value checking.

By using relative tolerance in assertEqual in unit tests, we should be able to simplify some of our complicated precision handling code described in issue #31135.


[ghstack-poisoned]

nairbv commented:
@gchanan  @ezyang @zou3519, et al... any thoughts on this change, or another approach that might be preferable?

Ideas:
1. we could have two arguments, aprec and rprec, and change all affected call sites. I'm not sure that's really preferable though since it seems like more hassle for callers.
2. we could add an optional `rprec=0.0` and just use it in particular cases that I want to refactor. This leaves most tests unaffected, though I think an end-state where we have a sensible non-zero default rtol would be preferable going forward.
3. we could add an rprec with a sensible default, and leave the existing absolute precision arg as-is? This is less invasive than option '1.' but still affects all tests and still complicates call sites that want to set precision.
4. We could just use the precision arg for both atol and rtol as in this PR, (and maybe alter the defaults in a subsequent PR).

Other ideas?

I'd hope to iterate on this. If we check dtypes of tensor arguments and use good defaults, I think we could eliminate many/most of the cases where we're currently explicitly specifying precision, and could also use tighter default tolerances. I may already have uncovered some bugs by playing around with this, where the wide berth we've set on atol (as high as 0.1) may have masked small correctness issues.

In general relative precision seems more useful, and seems like it would allow us to set tighter tolerances. 123456789.0f and 0.123456789f have the the same rtol but very different atol. If a tensor contains both values, an atol that covers the tolerance of the large value forces us to ignore meaningful errors in the small value. I don't think we can't _just_ use rtol either though since very small values can still be hard to represent accurately after simple operations like `tensor(1e-5) + 1 - 1`. Using both consistently seems ideal.

A change here potentially affects all existing tests so want some broader feedback.

zou3519 commented:
Some quick thoughts:

- I think that having the option of relative precision is nice. There are a lot of times when I am writing operator tests when my inputs are huge and the absolute precision of 1e-6 doesn't cut it. Sometimes I don't even realize the inputs to the tests are huge and a problem ends up manifesting down the line when someone changes the operator and are not sure if the precision decrease is OK or not.
- AFAICT this PR is changing assertEquals to do relative precision for everything, right? I don't think we should make a sweeping change in one go like this; It's possible that we have really low absolute tolerance (say 1) to test some operators that have outputs of around 1e7, and now changing that to relative tolerance would make the error bounds less strict.

nairbv commented:
> * I think that having the option of relative precision is nice. There are a lot of times when I am writing operator tests when my inputs are huge and the absolute precision of 1e-6 doesn't cut it. Sometimes I don't even realize the inputs to the tests are huge and a problem ends up manifesting down the line when someone changes the operator and are not sure if the precision decrease is OK or not.

I do feel like this could help correctness and consistency of a lot of tests long-term.

> * AFAICT this PR is changing assertEquals to do relative precision for everything, right? 

Right, as implemented now. There are a few other possible approaches I could go with, though some could be time consuming so don't want to go too far down a path we might not be happy with.

> I don't think we should make a sweeping change in one go like this; 

I do hope we can (eventually?) get to a state where the change is sweeping. Without changing the default, I'd think we don't encourage good settings going forward.

> It's possible that we have really low absolute tolerance (say 1) to test some operators that have outputs of around 1e7, and now changing that to relative tolerance would make the error bounds less strict.

This does universally make all error bounds somewhat less strict. Concretely, for any pair of values a, b, current code asserts `abs(a-b) < precision`. As-written here, this would instead assert `abs(a-b) < (precision + precision * b)`. When comparing small values, this should have very little impact. If someone used a wide `> 0` precision bound _because_ they were comparing large values, after this change their tolerance would be too high. I'd like to go through and find/fix most of those cases after this change. I'm not sure if it makes sense to try to do that all in one pass.

I do see cases in test_quantized where we have `prec=1.0`, so maybe as a minimum that should be fixed, but not sure if that's sufficient to land this.

Another extra careful approach could be to make a few passes: add an option with default 0.0, update tests to use a non-zero rtol and lower atol, reset the default, then remove the explicit precisions from call sites. That approach seems excessively cautious to me though and there's some value to the simplicity of a single precision parameter on `assertEquals`.

zou3519 commented:
numpy seems to recommend [this](https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) for their testing which has both relative and absolute tolerance arguments. I don't see the value in just having one argument -- both are useful in different situations

> This does universally make all error bounds somewhat less strict. Concretely, for any pair of values a, b, current code asserts abs(a-b) < precision. As-written here, this would instead assert abs(a-b) < (precision + precision * b). When comparing small values, this should have very little impact. If someone used a wide > 0 precision bound because they were comparing large values, after this change their tolerance would be too high. I'd like to go through and find/fix most of those cases after this change. I'm not sure if it makes sense to try to do that all in one pass. I do see cases in test_quantized where we have prec=1.0, so maybe as a minimum that should be fixed, but not sure if that's sufficient to land this.

We definitely shouldn't merge this change in as-is and then fix everything after. This change increases the degree of error allowed in our tests; merging this would weaken the tests

nairbv commented:
> numpy seems to recommend this for their testing which has both relative and absolute tolerance arguments. I don't see the value in just having one argument -- both are useful in different situations

I like that rtol comes first in theirs. I'd think relative would usually be the more important tolerance.

We could add a new function based on numpy and deprecate ours, though I think there's value in having  everything quickly move to rtol style tests.

> We definitely shouldn't merge this change in as-is and then fix everything after. This change increases the degree of error allowed in our tests; merging this would weaken the tests

```
[devgpu252.prn2 06 15:19:11 ~/repos/pytorch/test]$ gg -rn 'prec=' . | wc -l
83
```
I could go through and set somewhat tighter tolerances in these 83 locations if we're confident about what the API should look like. I'd be setting tighter tolerances somewhat arbitrarily though, not deeply analyzing the particular requirements of every call site.

ezyang commented:
I think it would be easier for us to come to an informed decision if we had some more information about how precise our current tests can be done by. Here is a simple experiment you could do to collect this data: for every call to `assertEquals` where we do a "test within error tolerance" check, print the absolute difference, and how much the relative difference would be (compared to the true thing). Put all of these in a big spreadsheet.

With this information, we'd be able to answer questions like:
1. How many tests had their error bounds widened under relative tolerance? Conversely, how many tests had their error bounds narrowed under relative tolerance?
2. How many tests could actually test for bit-for-bit equality?
3. How many tests have an absolute tolerance which translates into a too large relative tolerance for their range?

ezyang commented:
I added the numerical stability Quansight crew as reviewers to this PR, perhaps they have some more opinions on this matter.

gchanan commented:
I think Ed's suggestion is good.

My guess is that a NumPy-style separate rtol/atol is the right choice, but would be good to back up with numbers.

kurtamohler commented:
Based on the function that allclose() uses (∣input − other∣≤ atol + rtol ×∣other∣), we are checking if the difference is is less than an absolute tolerance plus the relative tolerance. 

In other words, an input could potentially fail both an individual absolute tolerance test and a relative tolerance test, but since we're summing the tolerances, it could pass the allclose() test. I'm wondering whether it would be preferable to check the absolute and relative differences separately by calling allclose twice, like this:

```
self.assertTrue(torch.allclose(a, b, atol=0, rtol=rprec, equal_nan=True), msg=message)
self.assertTrue(torch.allclose(a, b, atol=aprec, rtol=0, equal_nan=True), msg=message)
```

We check absolute and relative differences separately to ensure that both tolerances are met. Is there a reason we would not want them both to be checked separately, where we really do want to check if the sum of tolerances is met?

ezyang commented:
I agree with Kurt, we definitely shouldn't be applying tol to BOTH rtol and atol.

nairbv commented:
> Based on the function that allclose() uses (∣input − other∣≤ atol + rtol ×∣other∣), we are checking if the difference is is less than an absolute tolerance plus the relative tolerance.

That's right.
 
> ```
> self.assertTrue(torch.allclose(a, b, atol=0, rtol=rprec, equal_nan=True), msg=message)
> self.assertTrue(torch.allclose(a, b, atol=aprec, rtol=0, equal_nan=True), msg=message)
> ```
> 
> This checks absolute and relative differences separately to ensure that both tolerances are met. Is there a reason we would not want them both to be checked separately, where we really do want to check if the sum of tolerances is met?

I don't think this would be preferable, as it would require setting wide tolerances for _both_ atol and rtol, or putting more special casing throughout our testing code. My goal is to simplify testing code and reduce special casing.

E.g.:

```
>>> a=torch.tensor([1e-17, 1.23456789e20], dtype=torch.double)
>>> b= (a.to(torch.float)+1000 - 1000).to(torch.double)
>>> torch.allclose(a, b, atol=0.1, rtol=0)
False
>>> torch.allclose(a, b, atol=0, rtol=0.1)
False
>>> torch.allclose(a, b, atol=1e-5, rtol=1e-5)
True
>>>
```

Using both makes testing code much simpler, allows tighter tolerances, and works as-expected for most situations.

kurtamohler commented:
@nairbv , ah I see what you mean. Thanks for the example.

rgommers commented:
We went through a bunch of such assert functions in NumPy. `assert_allclose` is indeed considered the best of them at the moment. Having both `atol` and `rtol` is quite important. There's useful in different circumstances, so setting them to the same `prec` value doesn't seem like a nice thing to do long-term.

I do agree with @nairbv that testing `atol` and `rtol` separate isn't a good idea. They complement each other:
- `rtol` is most important
- for tensors with large scale variations, one would want `rtol=1e-X, atol=0`
- for tensors with zeros, one would want `rtol=1e-X, atol=1e-Y` (to avoid 0/0 issues)
- for tensors with both large scale variations and zeros, it's hard to pick a good setting, need to judge case by case and compromise a bit.

There's no setting that works universally. `rtol=1e-X, atol=0` as in numpy's `assert_allclose` is a decent default; it only breaks down when a test array includes zeros. For numpy-based code that only happens at test creation time, unless someone forgets to set a random seed. So at that point one needs to set a non-zero `atol` to guard against `0/0` errors. The PyTorch random functions aren't guaranteed to give the same bitstreams between platforms. That may be a reason to set a small nonzero `atol`. Something like `rtol=1e-5, atol=1e-7` could be reasonable.

nairbv commented:
[test_torch_tolerance.xlsx](https://github.com/pytorch/pytorch/files/4182844/test_torch_tolerance.xlsx)

Adding data based on logging allowable tolerances from test_torch.py. It's not all tests, but test_torch seems like a representative sample.

To get worst case tolerance when setting atol=rtol, I'm taking `abs_diff  <= prec + prec*other.abs()` and solving such that `prec = abs_diff / (1 + other.abs())`. There are some weird issues with memory-format tests (checking with @VitalyFedyunin), so the spreadsheet corrects the mixed precision to zero when the result is greater than max atol.

nairbv commented:
Specific example of the weird memory format test case that skews these results:
```
In [25]: x = torch.randn((4, 3, 8, 8 ), device='cpu').contiguous(memory_format=torch.channels_last)
In [26]: r1=x.sqrt()
In [27]: r2=x.contiguous().sqrt()
In [28]: diff=torch.abs(r1-r2)
# contains no non-zero non-nan values:
In [34]: diff.max()
Out[34]: tensor(nan)
In [35]: diff[diff==diff].max()
Out[35]: tensor(0.)
In [33]: diff[ (diff!=0) & (diff==diff) ]
Out[33]: tensor([])
# but this is non-zero ??
In [29]: (diff/(1+r2)).max()
Out[29]: tensor(1.7442e+28)
```

nairbv commented:
> Something like rtol=1e-5, atol=1e-7 could be reasonable.

Ideally I'd like to use defaults derived from the dtypes of the tensors being compared, based on worst-case from analysis of existing tests and/or a percentage of the bits in the significand of each floating point dtype.

nairbv commented:
This is replaced by https://github.com/pytorch/pytorch/pull/34258
